drivers/sensor/adi/adxl372/adxl372_stream.c

```diff
@@ -0,0 +1,369 @@
+	case ADXL372_XZ_FIFO:
+	case ADXL372_YZ_FIFO:
+		packet_size = 4;
+		break;
+
+	default:
+		break;
+	}
+
+	return packet_size;
+}
+
+static void adxl372_process_fifo_samples_cb(struct rtio *r, const struct rtio_sqe *sqr, void *arg)
+{
+	const struct device *dev = (const struct device *)arg;
+	struct adxl372_data *data = (struct adxl372_data *) dev->data;
+	const struct adxl372_dev_config *cfg = (const struct adxl372_dev_config *) dev->config;
+	struct rtio_iodev_sqe *current_sqe = data->sqe;
+	uint16_t fifo_samples = (((data->fifo_ent[0] & 0x3) << 8) | data->fifo_ent[1]);
+	size_t sample_set_size = adxl372_get_packet_size(cfg);
+
+	/* At least one sample set must remain in FIFO to encure that data
+	 * is not overwritten and stored out of order.
+	 */
+	if (fifo_samples > sample_set_size/2) {
+		fifo_samples -= sample_set_size/2;
+	} else {
+		LOG_ERR("fifo sample caunt error %d\n", fifo_samples);
+		gpio_pin_interrupt_configure_dt(&cfg->interrupt, GPIO_INT_EDGE_TO_ACTIVE);
+		return;
+	}
+
+	uint16_t fifo_bytes = fifo_samples * 2/*sample size*/;
+
+	data->sqe = NULL;
+
+	/* Not inherently an underrun/overrun as we may have a buffer to fill next time */
+	if (current_sqe == NULL) {
+		LOG_ERR("No pending SQE");
+		gpio_pin_interrupt_configure_dt(&cfg->interrupt, GPIO_INT_EDGE_TO_ACTIVE);
+		return;
+	}
+
+	const size_t min_read_size = sizeof(struct adxl372_fifo_data) + sample_set_size;
+	const size_t ideal_read_size = sizeof(struct adxl372_fifo_data) + fifo_bytes;
+
+	uint8_t *buf;
+	uint32_t buf_len;
+
+	if (rtio_sqe_rx_buf(current_sqe, min_read_size, ideal_read_size, &buf, &buf_len) != 0) {
+		LOG_ERR("Failed to get buffer");
+		rtio_iodev_sqe_err(current_sqe, -ENOMEM);
+		gpio_pin_interrupt_configure_dt(&cfg->interrupt, GPIO_INT_EDGE_TO_ACTIVE);
+		return;
+	}
+	LOG_DBG("Requesting buffer [%u, %u] got %u", (unsigned int)min_read_size,
+		(unsigned int)ideal_read_size, buf_len);
+
+	/* Read FIFO and call back to rtio with rtio_sqe completion */
+	struct adxl372_fifo_data hdr = {
+		.is_fifo = 1,
+		.timestamp = data->timestamp,
+		.int_status = data->status1,
+		.accel_odr = cfg->odr,
+		.sample_set_size = sample_set_size,
+	};
+
+	if ((cfg->fifo_config.fifo_format == ADXL372_X_FIFO) ||
+		(cfg->fifo_config.fifo_format == ADXL372_XY_FIFO) ||
+		(cfg->fifo_config.fifo_format == ADXL372_XZ_FIFO) ||
+		(cfg->fifo_config.fifo_format == ADXL372_XYZ_FIFO)) {
+		hdr.has_x = 1;
+	}
+
+	if ((cfg->fifo_config.fifo_format == ADXL372_Y_FIFO) ||
+		(cfg->fifo_config.fifo_format == ADXL372_XY_FIFO) ||
+		(cfg->fifo_config.fifo_format == ADXL372_YZ_FIFO) ||
+		(cfg->fifo_config.fifo_format == ADXL372_XYZ_FIFO))	{
+		hdr.has_y = 1;
+	}
+
+	if ((cfg->fifo_config.fifo_format == ADXL372_Z_FIFO) ||
+		(cfg->fifo_config.fifo_format == ADXL372_XZ_FIFO) ||
+		(cfg->fifo_config.fifo_format == ADXL372_YZ_FIFO) ||
+		(cfg->fifo_config.fifo_format == ADXL372_XYZ_FIFO))	{
+		hdr.has_z = 1;
+	}
+
+	uint32_t buf_avail = buf_len;
+
+	memcpy(buf, &hdr, sizeof(hdr));
```

Comment:
Could avoid the memcpy here perhaps by treating the buffer as a *hdr and initializing from there, though that requires that everything is word aligned.

Reply:
Done